src/components/KTable/ColumnVisibilityMenu.vue

```diff
@@ -0,0 +1,91 @@
+<template>
+  <div class="k-table-column-visibility-menu">
+    <KDropdown>
+      <KTooltip text="Show/Hide Columns">
+        <KButton
+          appearance="secondary"
+          class="menu-button"
+        >
+          <template #icon>
+            <CogIcon />
+          </template>
+        </KButton>
+      </KTooltip>
+
+      <template #items>
+        <KDropdownItem
+          v-for="col in columns"
+          :key="col.key"
+          class="column-visibility-menu-item"
+          @click.stop="visibilityMap[col.label] = !visibilityMap[col.label]"
+        >
+          <KCheckbox
+            v-model="visibilityMap[col.label]"
+            :label="col.label"
+          />
+        </KDropdownItem>
+        <KDropdownItem>
+          <KButton
+            appearance="tertiary"
+            class="apply-button"
+            @click="handleApply"
+          >
+            Apply
+          </KButton>
+        </KDropdownItem>
+      </template>
+    </KDropdown>
+  </div>
+</template>
+
+<script setup lang="ts">
+import { ref, onBeforeMount, type PropType } from 'vue'
+import type { TableHeader } from '@/types'
+import { CogIcon } from '@kong/icons'
+import KButton from '@/components/KButton/KButton.vue'
+import KCheckbox from '@/components/KCheckbox/KCheckbox.vue'
+import KDropdown from '@/components/KDropdown/KDropdown.vue'
+import KDropdownItem from '@/components/KDropdown/KDropdownItem.vue'
+
+const emit = defineEmits<{
+  (e: 'update:visibility', columnVisibility: Record<string, boolean>): void
+}>()
+
+const props = defineProps({
+  columns: {
+    type: Array as PropType<TableHeader[]>,
+    required: true,
+  },
+  visibilityPreferences: {
+    type: Object as PropType<Record<string, boolean>>,
+    default: () => ({}),
+  },
+})
+
+const visibilityMap = ref<Record<string, boolean>>({})
+
+const handleApply = () => {
+  // pass by ref problems
+  emit('update:visibility', JSON.parse(JSON.stringify(visibilityMap.value)))
```

Comment:
question: why do we need `JSON.parse(JSON.stringify(` here? Can't we just emmit the original object, or does that produce mutations?